design/operater-gateway-support-migration-design.md

```diff
@@ -0,0 +1,88 @@
+  - Checks if named GatewayClass has the correct (hardcoded to `projectcontour.io/contour-operator`) controller string
+  - Checks for Contour CRD in the parameters ref of the GatewayClass, uses parameters for provisioning Contour instances
+  - Sets admitted status on object if all valid
+- Includes Gateway controller
+  - Checks if Gateway refers to correct GatewayClass and that GatewayClass has been admitted
+  - Sets up finalizer for Gateway
+  - Creates Contour deployment resources with the Gateway being reconciled named in Contour's ConfigMap
+  - Sets up finalizer for GatewayClass
+  - Sets up finalizer for Contour CRD referenced in GatewayClass
+  - Sets admitted status on object if all valid
+
+### Contour Controller
+- Needs to be configured with a GatewayClass controller name to ensure it reconciles the correct GatewayClass
+- Also needs a Gateway namespace/name which is being deprecated
+- Includes GatewayClass controller
+  - Expects parameter ref field to be unset, otherwise GatewayClass is not admitted
+  - Admits GatewayClass that has the configured controller name
+- Includes Gateway controller
+  - Checks if Gateway references Contour owned GatewayClass
+  - Passes off reconciliation to dag package Gateway API processor
+  - Gateway is validated, status set on it and relevant *Routes
+
+### Current Problems
+- Contour requires Gateway namespace/name *and* controller name, the latter of which the operator does not provide
+- Contour expects GatewayClass parameters ref to be unset while Operator expects it to reference a Contour CRD
+- If the above is rectified, Contour and the Operator still will conflict on setting status
+- CI failures that occur because of conflicts between Contour and the Operator are not detected unless we manually look at the scheduled Operator CI job
+
+## Goals
+- Improve implementation in Contour and Operator to mitigate current issues
+- Improve development processes and practices to ensure this conflicting state does not occur again
+
+## Non Goals
+- Drop support for features or projects
+
+## Remediation Steps
+
+- Set up alerting on Operator CI failures
+  - So we catch issues sooner.
+- Operator sets controller name in Contour ConfigMap
+  - To ensure Contour does not immediately exit on startup.
+- Operator stops setting any status on Gateway API resources
+  - To ensure Contour and Operator do not "fight" over status
+- Contour does not reject GatewayClasses if parameters ref is set
+  - At this point, we should have a workable intermediate solution for the moment and can cut a patch release
+  - Reconciliation will be duplicated but this should "work" for now
+  - User creates GatewayClass/Contour CRD with GatewayClass name
+  - Operator reconciles, does not set status
+  - User creates Gateway referencing GatewayClass
+  - Operator reconciles, creates a Contour
+  - Contour reconciles GatewayClass and Gateway, sets status, etc.
+- Operator deprecates Contour CRD GatewayClassRef field
```

Comment:
do you mean we're deprecating the spec.gatewayClassRef from the Contour CR entirely

Reply:
Originally we were using name/namespace of a Gateway. Then the operator moved to reference a GatewayClass.Name. Finally we're at using a GatewayClass.ControllerName and looking up the GatewayClass.Name matching a Gateway.Spec.ClassName.